nilearn/mass_univariate/permuted_least_squares.py

```diff
@@ -611,18 +614,290 @@ def permuted_ols(
+            "Some descriptors in 'target_vars' have zeros across all samples. "
+            "These descriptors will be ignored during null distribution "
+            "generation."
+        )
+
+    tested_var_has_intercept = _check_for_intercept_in_tested_var(tested_vars)
+
+    (
+        confounding_vars,
+        intercept_test,
+        model_intercept,
+    ) = _check_for_intercept_in_confounds(
+        confounding_vars, tested_var_has_intercept, model_intercept
+    )
+
+    confounding_vars = _add_intercept_optionally(
+        confounding_vars,
+        model_intercept,
+        intercept_test,
+        n_samples=tested_vars.shape[0],
+    )
+
+    # OLS regression on original data
+    # original regression (= regression on residuals + adjust t-score)
+    # compute t score map of each tested var for original data
+    # scores_original_data is in samples-by-regressors shape
+    confounding_vars = _orthonormalize_confounding_vars(confounding_vars)
+    target_vars = _prepare_target_vars(target_vars, confounding_vars)
+    tested_vars = _prepare_tested_vars(tested_vars, confounding_vars)
+    scores_original_data = _t_score_with_covars_and_normalized_design(
+        tested_vars,
+        target_vars.T,
+        confounding_vars,
+    )
+
+    # Define connectivity for TFCE and/or cluster measures
+    tfce_original_data = _get_tfce_original_data(
+        tfce, masker, scores_original_data, two_sided_test
+    )
+
+    # 0 or negative number of permutations => original data scores only
+    if n_perm <= 0:
+        if output_type == "legacy":
+            return np.asarray([]), scores_original_data.T, np.asarray([])
+
+        out = {"t": scores_original_data.T}
+
+        if tfce:
+            out["tfce"] = tfce_original_data.T
+
+        return out
+
+    # Permutations
+    # parallel computing units perform a reduced number of permutations each
+    return _run_permutations(
+        tested_vars,
+        target_vars,
+        confounding_vars,
+        n_perm,
+        random_state,
+        two_sided_test,
+        n_jobs,
+        verbose,
+        masker,
+        tfce,
+        threshold,
+        output_type,
+        n_descriptors,
+        intercept_test,
+        scores_original_data,
+        tfce_original_data,
+    )
+
+
+def _run_permutations(
+    tested_vars,
+    target_vars,
+    confounding_vars,
+    n_perm,
+    random_state,
+    two_sided_test,
+    n_jobs,
+    verbose,
+    masker,
+    tfce,
+    threshold,
+    output_type,
+    n_descriptors,
+    intercept_test,
+    scores_original_data,
+    tfce_original_data,
+):
+    """Run permutations and compute p-values.
+
+    If permutation test was requested, this function will run
+    the n_perm permutations and compute the p-values of
+    the voxel-wise scores_original_data
+    and the p-values of the cluster-wise tfce_original_data
+    and return the proper output_type.
+
+    See permuted_ols for a description of the parameters.
+    """
```

Comment:
added some big overview of what the function does.

Note: if anyone has good tips on how to minimize copy-pasting parameters definition a bazillion times across docstrings within the same module, let me know.